### PR TITLE
Suggestion: Replace PySide with QtPy

### PR DESCRIPTION
**Pull request by AndrewAmmerlaan**
Might I suggest to use https://github.com/spyder-ide/qtpy instead of importing PySide2 directly. QtPy is a abstraction layer for PyQt(5) and PySide(2), using this instead of explicitly choosing for PySide2 would increase compatibility with distros that only have one of the two. And it would also prevent users having to install both PyQt5 and PySide2. It is my understanding that PySide2 and PyQt5 are almost identical anyway, and it should be possible to just replace all import PySide.foobar with import qtpy.foobar without breaking anything. I have done just that in this PR, and it seems to work just fine for me, what do you think?

### CODE REVIEW

File: vidify/__main__.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ def start_gui(config: Config) -> None:
     is put together.
     """
 
-    app = QApplication()
+    app = QApplication(sys.argv)
```
Comment (marioortizmanero) [inline vidify/__main__.py]: This isn't necessary. I already get the arguments and config in the `config` module.

File: dev/linux.spec [hunk 2]
```diff
@@ -12,7 +12,7 @@ a = Analysis(['../vidify/__main__.py'],
                     ('/usr/lib/vlc/libvlc_pulse.so', '.'),
                     ('/usr/lib/vlc/libvlc_vdpau.so', '.'),
              ],
-             hiddenimports=['appdirs', 'PySide2', 'vlc', 'mpv', 'packaging',
+             hiddenimports=['appdirs', 'qtpy', 'vlc', 'mpv', 'packaging',
```
Comment (marioortizmanero) [inline dev/linux.spec]: See the main comment

File: dev/linux_onefile.spec [hunk 3]
```diff
@@ -12,7 +12,7 @@ a = Analysis(['../vidify/__main__.py'],
                     ('/usr/lib/vlc/libvlc_pulse.so', '.'),
                     ('/usr/lib/vlc/libvlc_vdpau.so', '.'),
              ],
-             hiddenimports=['appdirs', 'PySide2', 'vlc', 'mpv', 'packaging',
+             hiddenimports=['appdirs', 'qtpy', 'vlc', 'mpv', 'packaging',
```
Comment (marioortizmanero) [inline dev/linux_onefile.spec]: See the main comment

File: setup.py [hunk 4]
```diff
@@ -33,7 +33,7 @@
     keywords='spotify music video player videos lyrics linux windows macos',
     python_requires='>=3.7',
     install_requires=[
-        'PySide2',
+        'QtPy',
```
Comment (marioortizmanero) [inline setup.py]: see the main comment

File: vidify/gui/api/spotify_web.py [hunk 5]
```diff
@@ -4,9 +4,9 @@
 
 import logging
 
-from PySide2.QtWidgets import (QWidget, QLabel, QPushButton, QHBoxLayout,
+from qtpy.QtWidgets import (QWidget, QLabel, QPushButton, QHBoxLayout,
                                QVBoxLayout)
```
Comment (marioortizmanero) [inline vidify/gui/api/spotify_web.py]: ./vidify/gui/api/spotify_web.py:8:32: E127 continuation line over-indented for visual indent

you need to remove a couple spaces to fix the spacing

File: vidify/gui/components.py [hunk 6]
```diff
@@ -6,12 +6,12 @@
 import logging
 from typing import Optional
 
-from PySide2.QtWidgets import (QWidget, QLabel, QPushButton, QLineEdit,
+from qtpy.QtWidgets import (QWidget, QLabel, QPushButton, QLineEdit,
                                QVBoxLayout, QGridLayout, QGroupBox)
```
Comment (marioortizmanero) [inline vidify/gui/components.py]: ./vidify/gui/components.py:10:32: E127 continuation line over-indented for visual indent

you need to remove a couple spaces here too

File: vidify/__main__.py (hunk 1 above)
Comment (AndrewAmmerlaan) [reply vidify/__main__.py]: I get this without it:
```
Traceback (most recent call last):
File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
"__main__", mod_spec)
File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
exec(code, run_globals)
File "vidify/__main__.py", line 58, in <module>
main()
File "vidify/__main__.py", line 54, in main
start_gui(config)
File "vidify/__main__.py", line 26, in start_gui
app = QApplication()
TypeError: QApplication(List[str]): not enough arguments
```

**Comment by AndrewAmmerlaan** [comment]
> For what I've seen though, QtPy is just the abstraction layer, so I'd need to list PyQt5 or PySide2 too in the setup.py and build dependencies, right? Or how does this work? You can see the Travis CI build errors yourself.

From the QtPy README file:
> ### Requirements
>
>You need PyQt5, PyQt4, PySide2 or PySide installed in your system to make use of QtPy. If several of these packages are found, PyQt5 is used by default unless you set the `QT_API` environment variable.
>
>`QT_API` can take the following values:
>
>-    pyqt5 (to use PyQt5).
>-    pyqt or pyqt4 (to use PyQt4).
>-    pyside2 (to use PySide2)
>-    pyside (to use PySide).

The beauty of the thing is that you only need to specify QtPy as dependency, the user can then choose to use either PyQt5 or PySide2. This is useful because PyQt5 and PySide2 are basically the same thing, and it is a bit of a waste to have to install both of them.

File: vidify/__main__.py (hunk 1 above)
Comment (marioortizmanero) [reply vidify/__main__.py]: Ah, that can be fixed by using `QApplication([])` then. I had seen it like this in tutorials but didn't know it was only for PyQt5.

Comment (AndrewAmmerlaan) [reply vidify/__main__.py]: > Ah, that can be fixed by using `QApplication([])` then. I had seen it like this in tutorials but didn't know it was only for PyQt5.

That gives me:
```
QSettings::value: Empty key passed
QSettings::value: Empty key passed
```

Comment (marioortizmanero) [reply vidify/__main__.py]: How about `QApplication(['vidify'])`? `vidify` there acts as the `argv[0]` and no other arguments passed.

**Comment by marioortizmanero** [comment]
> I see the Travis CI errors, and I think they are due to the fact that QtPy's setup.py does not explicitly list PySide or PyQt as a dependency. These dependencies are usually handled by the distribution's own package manager, e.g. on Gentoo:

Seems like on the [AUR](https://aur.archlinux.org/packages/python-qtpy-git/) it simply installs PyQt5 directly. Also, it's a `-git` version so I might have to upload a regular `python-qtpy` (also because it's outdated). But it's the best solution so I'll take it.

> Because neither PyQt5 or PySide are explicitly required by QtPy the Travis CI bot does not install either and therefore fails to find any Qt bindings. I'm not sure how to fix that though, maybe there is a way to get the Travis CI bot to install PySide2 or PyQt5 without explicitly listing these dependencies in vidivy's setup.py file. Because that would defeat the whole purpose of using QtPy in the first place.

I know what you're saying but I wanted this to be automatic without the user knowing. Your solution in the `setup.py` is pretty neat, thanks!

But be careful because a user could only have `PyQt5` installed without the `PyQtWebEngine` and your `setup.py` would not install the latter.

And it does work for me:

```
>>> get_dist('PyQt5')
PyQt5 5.14.0 (/home/mario/Programming/vidify/lib/python3.8/site-packages)
>>> get_dist('PySide2')
PySide2 5.14.0 (/home/mario/Programming/vidify/lib/python3.8/site-packages)
```

At least inside my virtual environment.

File: vidify/__main__.py (hunk 1 above)
Comment (AndrewAmmerlaan) [reply vidify/__main__.py]: > How about `QApplication(['vidify'])`? `vidify` there acts as the `argv[0]` and no other arguments passed.

That works :smile:

**Comment by AndrewAmmerlaan** [comment]
> Seems like on the AUR it simply installs PyQt5 directly. Also, it's a -git version so I might have to upload a regular python-qtpy (also because it's outdated). But it's the best solution so I'll take it.

There's also [this](https://www.archlinux.org/packages/community/any/python-qtpy/) in Arch's Community repo, which is the most recent version.

**Comment by AndrewAmmerlaan** [comment]
> But be careful because a user could only have PyQt5 installed without the PyQtWebEngine and your setup.py would not install the latter. Also, if the user has PySide or PyQt4 installed, this wouldn't work. Is this because there are compatibility issues with these two?

Fixed the problem with PyQtWebEninge. Regarding, Pyside and PyQt4 these are Qt4 packages and I think they are no longer used (at least not in Gentoo), but I suppose it can't hurt to add those to the `if` statement (unless vidify doesn't work with Qt4, I can't test that because Qt4 no longer is in the Gentoo repo)

**Comment by AndrewAmmerlaan** [comment]
> EDIT: oh wait if I install it from the Arch repos it doesn't work:
>
>>>> get_dist('PySide2')
>>>> get_dist('PyQt5')
>PyQt5 5.14.0 (/usr/lib/python3.8/site-packages)

>I can open an issue in https://github.com/spyder-ide/qtpy, they'll know better in there.

>EDIT2: I should also modify Travis CI to test the app with both PySide2 and PyQt5 to make sure I don't break one of them

I suspect the Travis CI bot installs it using pip? Maybe that package does include the `.egg-info` directory that contains the necessary metadata. In any case this is not a bug in QtPy, it's a problem with the build system of PySide2. If you take a look at the setup.py file in the PySide2 release tarball you'll find that it doesn't have a proper `setup()` function with metadata. Therefore when running setuptools on that setup.py file it won't be able to create a `.egg-info` directory with metadata (I think).

**Comment by marioortizmanero** [comment]
> There's also this in Arch's Community repo, which is the most recent version.

Whoops I forgot to check the official repos haha. Anyway, this package does seem to not install any of the Qt bindings. It just leaves them as optional. So I'll have to choose one of them as mandatory in my `PKGBUILD` anyway so that it's automatic (PKGBUILDS are supposed to be simple).

> Fixed the problem with PyQtWebEninge. Regarding Pyside and PyQt4, these are Qt4 packages and I think they are no longer used (at least not in Gentoo), but I suppose it can't hurt to add those to the if statement (unless vidify doesn't work with Qt4, I can't test that because Qt4 got removed from the Gentoo repo)

Sorry! I just realized that `pyside` is only available for versions that `vidify` doesn't support:

```
    only these python versions are supported: [(2, 6), (2, 7), (3, 2), (3, 3), (3, 4)]
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

and that `pyqt4` isn't even on pip. Please remove them back.

**Comment by AndrewAmmerlaan** [comment]
> Please remove them back.

Fixed

**Comment by marioortizmanero** [comment]
Thanks! I'm merging the PR then :)

**Comment by marioortizmanero** [comment]
Argh I'm getting an error now when I use PySide2:

```
Traceback (most recent call last):
  File "/home/mario/Programming/vidify/lib/python3.8/site-packages/qtpy/QtWebEngineWidgets.py", line 22, in <module>
    from PyQt5.QtWebEngineWidgets import QWebEnginePage
ModuleNotFoundError: No module named 'PyQt5.QtWebEngineWidgets'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/mario/Programming/vidify/vidify/__main__.py", line 16, in <module>
    from vidify.gui.window import MainWindow
  File "/home/mario/Programming/vidify/vidify/gui/window.py", line 27, in <module>
    from vidify.gui.components import APISelection
  File "/home/mario/Programming/vidify/vidify/gui/components.py", line 14, in <module>
    from qtpy.QtWebEngineWidgets import QWebEngineView
  File "/home/mario/Programming/vidify/lib/python3.8/site-packages/qtpy/QtWebEngineWidgets.py", line 26, in <module>
    from PyQt5.QtWebKitWidgets import QWebPage as QWebEnginePage
ModuleNotFoundError: No module named 'PyQt5.QtWebKitWidgets'
```

**Comment by AndrewAmmerlaan** [comment]
> Argh I'm getting an error now when I use PySide2:
> 

Does it work if you do: `QT_API=PySide2 python vidify`?

**Comment by marioortizmanero** [comment]
Yes, but I thought it detected PySide automatically?

**Comment by marioortizmanero** [comment]
Oh yes my bad. I forgot to run an install before running vidify. Fixed!

**Comment by AndrewAmmerlaan** [comment]
> Yes, but I thought it detected PySide automatically?

It should, but it doesn't use it if PyQt5 is also present, because then it defaults to PyQt5

**Pull request merged**